src/shell/shell_cmds.cpp

```diff
@@ -2242,3 +2243,271 @@ void DOS_Shell::CMD_VOL(char* args)
+			quote = !quote;
+			memmove(args, args + 1, strlen(args));
+			break;
+		case ' ':
+			if (quote) {
+				args++;
+			} else {
+				*args++ = 0;
+				parsing = false;
+			}
+			break;
+		case '\0': parsing = false; break;
+		default: args++;
+		}
+	}
+
+	if (!*raw_destination) {
+		WriteOut(MSG_Get("SHELL_MISSING_PARAMETER"));
+		return;
+	}
+
+	// Check for extra arguments
+	while (*args) {
+		if (*args++ != ' ') {
+			WriteOut(MSG_Get("SHELL_TOO_MANY_PARAMETERS"));
+			return;
+		}
+	}
+
+	char full_destination[DOS_PATHLENGTH];
+	if (!DOS_Canonicalize(raw_destination, full_destination)) {
+		WriteOut(MSG_Get("SHELL_FILE_NOT_FOUND"), raw_destination);
+		return;
+	}
+
+	// Command uses dta so set it to our internal dta
+	const RealPt save_dta = dos.dta();
+	dos.dta(dos.tables.tempdta);
+	DOS_DTA dta(dos.dta());
+
+	// Each raw source can be a wildcard (*.*) containing multiple real
+	// sources.
+	std::vector<std::string> final_sources;
+	for (size_t i = 0; i < num_sources; ++i) {
+		if (!DOS_FindFirst(raw_sources[i], 0xffff & ~DOS_ATTR_VOLUME)) {
+			WriteOut(MSG_Get("SHELL_FILE_NOT_FOUND"), raw_sources[i]);
+			continue;
+		}
+		char source_dir[DOS_PATHLENGTH];
+		if (!DOS_Canonicalize(raw_sources[i], source_dir)) {
+			WriteOut(MSG_Get("SHELL_FILE_NOT_FOUND"), raw_sources[i]);
+			continue;
+		}
+		*(strrchr(source_dir, '\\') + 1) = 0;
+		do {
+			DOS_DTA::Result result = {};
+			dta.GetResult(result);
+			if (result.name != "." && result.name != "..") {
+				final_sources.emplace_back(source_dir + result.name);
+			}
+		} while (DOS_FindNext());
+	}
+
+	// Done with DTA. Restore it.
+	dos.dta(save_dta);
+
+	uint16_t destination_attr = 0;
+	bool destination_exists = DOS_GetFileAttr(full_destination, &destination_attr);
+	size_t destination_len = strlen(full_destination);
+	bool dest_is_dir       = false;
+
+	if (destination_exists) {
+		if (destination_attr & DOS_ATTR_DIRECTORY) {
+			dest_is_dir = true;
+		} else if (final_sources.size() > 1) {
+			WriteOut(MSG_Get("SHELL_CMD_MOVE_MULTIPLE_TO_SINGLE"));
+			return;
+		}
+	} else if (final_sources.size() > 1) {
+		dest_is_dir = true;
+		if (!DOS_MakeDir(full_destination)) {
+			WriteOut(MSG_Get("SHELL_CMD_MKDIR_ERROR"), full_destination);
+			return;
+		}
+	}
+
+	if (dest_is_dir && full_destination[destination_len - 1] != '\\') {
+		if (destination_len + 1 >= ARRAY_LEN(full_destination)) {
+			WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+			return;
+		}
+		full_destination[destination_len++] = '\\';
+		full_destination[destination_len]   = 0;
+	}
+
+	for (const std::string& source_str : final_sources) {
+		const char* source_c = source_str.c_str();
+		if (dest_is_dir) {
+			const char* filename = strrchr(source_c, '\\') + 1;
+			size_t filename_len  = strlen(filename);
+			if (filename_len + destination_len >=
+			    ARRAY_LEN(full_destination)) {
+				WriteOut(MSG_Get("SHELL_ILLEGAL_PATH"));
+				continue;
+			}
+			memcpy(full_destination + destination_len,
+			       filename,
```

Comment:
The ideal number of C string functions in new code is exactly zero 😄 Especially for non-performance intensive code. `const char*` is fine for string literals though that don't get manipulated.

Could you rewrite this using std::string only?

Reply:
Fixed